ArduCopter/mode_auto.cpp

```diff
@@ -324,32 +324,32 @@ void ModeAuto::takeoff_start(const Location& dest_loc)
     // by default current_alt_cm and alt_target_cm are alt-above-EKF-origin
     int32_t alt_target_cm;
     bool alt_target_terrain = false;
-    float current_alt_cm = inertial_nav.get_position_z_up_cm();
+    int32_t current_alt_cm = static_cast<int32_t>(inertial_nav.get_position_z_up_cm());
     float terrain_offset;   // terrain's altitude in cm above the ekf origin
-    if ((dest_loc.get_alt_frame() == Location::AltFrame::ABOVE_TERRAIN) && wp_nav->get_terrain_offset(terrain_offset)) {
+    // Record the takeoff target and set the location to current position as we always takeoff from current position.
+    // we use now takeoff_loc instead of dest_loc to ensure we are getting altitude offset from the current position and not whatever inside the mission command
+    takeoff_loc = Location(copter.current_loc.lat, copter.current_loc.lng, dest_loc.alt, dest_loc.get_alt_frame());
+
+    if ((takeoff_loc.get_alt_frame() == Location::AltFrame::ABOVE_TERRAIN) && wp_nav->get_terrain_offset(terrain_offset)) {
         // subtract terrain offset to convert vehicle's alt-above-ekf-origin to alt-above-terrain
-        current_alt_cm -= terrain_offset;
+        current_alt_cm -= static_cast<int32_t>(terrain_offset);
```

Comment:
I don't think we need this cast do we?

Reply:
needed to save bytes as current_alt_cm is int now.

Reply:
keep as float please.